floss/identify.py

```diff
@@ -0,0 +1,130 @@
+    calls_to = set()
+
+    for fva in vw.getFunctions():
+        if skip_thunks and is_thunk_function(vw, fva):
+            continue
+
+        # TODO skip_libs and is_library_function
+        #     continue
+
+        calls_to.add(len(vw.getXrefsTo(fva)))
+
+    return max(calls_to)
+
+
+def is_thunk_function(vw, function_address):
+    return vw.getFunctionMetaDict(function_address).get("Thunk", False)
+
+
+def get_function_score_weighted(features):
+    return sum(feature.weighted_score() for feature in features) / sum(feature.weight for feature in features)
+
+
+def find_decoding_functions(vw, functions, disable_progress=None):
+    decoding_candidate_functions = collections.defaultdict(float)
+
+    pbar = tqdm.tqdm
+    if disable_progress:
+        # do not use tqdm to avoid unnecessary side effects when caller intends
+        # to disable progress completely
+        pbar = lambda s, *args, **kwargs: s
+
+    # TODO see FLIRT below
+    # n_funcs = len(functions)
+
+    functions = sorted(functions)
+
+    pb = pbar(functions, desc="finding decoding functions", unit=" functions", postfix="skipped 0 library functions")
+    for f in pb:
+        function_address = int(f)
+
+        if is_thunk_function(vw, function_address):
+            continue
+
+        # TODO ignore library functions via FLIRT
+        # if viv_utils.flirt.is_library_function(vw, function_address):
+        #     function_name = viv_utils.get_function_name(vw, function_address)
+        #     logger.debug("skipping library function 0x%x (%s)", function_address, function_name)
+        #     meta["library_functions"][function_address] = function_name
+        #     n_libs = len(meta["library_functions"])
+        #     percentage = 100 * (n_libs / n_funcs)
+        #     if isinstance(pb, tqdm.tqdm):
+        #         pb.set_postfix_str("skipped %d library functions (%d%%)" % (n_libs, percentage))
+        #     continue
+
+        f = viv_utils.Function(vw, function_address)
+
+        function_data = {"meta": get_function_meta(f), "features": list()}
+
+        # meta data features
+        function_data["features"].append(BlockCount(function_data["meta"].get("block_count")))
+        function_data["features"].append(InstructionCount(function_data["meta"].get("instruction_count")))
+        function_data["features"].append(Arguments(function_data["meta"].get("api", []).get("arguments")))
+
+        for feature in extract_function_features(f):
+            function_data["features"].append(feature)
+
+        for bb in f.basic_blocks:
+            for feature in extract_basic_block_features(f, bb):
+                function_data["features"].append(feature)
+
+            for insn in bb.instructions:
+                for feature in extract_insn_features(f, bb, insn):
+                    function_data["features"].append(feature)
+
+        for feature in abstract_features(function_data["features"]):
+            function_data["features"].append(feature)
```

Comment:
to simplify this we could just extract really relevant features (e.g. nzxor, most called functions) or their combinations and target all functions that have at least one match

Reply:
this is probably a good idea to do less work